### ISSUE TITLE
Slackbot router: design brainstorming

### ISSUE THREAD

**Issue by ramyaragupathy** [hotosm/slack-bots#2]
For this summer, @andria-hibe will be working with us to setup a router project that takes in multiple triggers through user-commands and external events and route it to respective micro service on AWS, which in turn post the results to slack.

We've started this document to capture use cases, router design and tool/platform selection. https://docs.google.com/document/d/13R8MUV3m6iKpL_aoJ5V6DEXuRsHVN5S4bz4kfFBhx3M/edit#heading=h.jcdx6rvcnxep

**Comment by andria-hibe** [comment]
Putting @arunasank architecture diagram and proposal here so we can have all our ideas discussions in one place:
---
![image](https://user-images.githubusercontent.com/3166852/82841811-6e90d780-9ea5-11ea-821c-94efc83038ef.png)


## Architecture Details

**Router**: An _AWS API Gateway_ module that has two end-points, `/slack-to-lambda` and `/lambda-to-slack`. The first end point is triggered by Slack, when a _slash-command_ is used on Slack. The second end point is triggered by a Lambda function once the results of a request have been processed.

**Lambda-SNS:** Each SNS-Lambda combination is named for the command. So, if the slash-command is called `tm-stats`, the SNS topic is named `tm-stats` and the lambda function is also named `tm-stats`. The slash-command `tm-stats` should only trigger the `tm-stats` SNS and thus, the `tm-stats` lambda. 

## Advantages
* **Logical separation** - Router is separated from the lambda functions. Once the router is completed, it need not be reworked unless we want to include a new input and output channel, say Github.
* **Modular architecture** - Each SNS-Lambda pair constitutes one command. Each command has its own Github repository, deploy template and AWS resources. AWS resources for one Slack command are not dependent on the AWS resources for another Slack command. A new command is integrated by:
  * Creating a new repository on GitHub
  * Creating associated SNS and Lambda functions and log resources using a deployment template
  * Creating scripts for the Lambda.

Separating out the Lambda functions makes it easy to do the following:
  * **Debugging and troubleshooting** - Routing issues can be debugged on the routing repository/router logs. Command related issues can be debugged in the command repository/command logs.
  * **Improved accessibility for new contributors** - Smaller repositories means smaller code-bases. New contributors can tackle smaller and simpler code-bases, having a larger chance of understanding the codebase without much intervention from HOT maintainers.
  * **Separation of secrets** - Secrets associated with one command/repository is only part of the deployment template of that command. Slack or channel related secrets are only part of the router repository.
  * **Improved maintainability** - Adding a new command will simply involve creating a new GitHub repository. Since the deployment templates of all the lambda functions involve the SNS, Lambda, Logs combination, it is easy to create a new deployment template. Maybe we can even write a template generator as part of this summer.

**Comment by andria-hibe** [comment]
Thanks for the diagram and the explanation @arunasank!

I have a couple of questions so I can understand better:

1. > **Router**: An _AWS API Gateway_ module that has two end-points, `/slack-to-lambda` and `/lambda-to-slack`. The first end point is triggered by Slack, when a _slash-command_ is used on Slack. The second end point is triggered by a Lambda function once the results of a request have been processed.

Regarding the /lambda-to-slack endpoint, I’m not sure how this works. Is this calling another Lambda function that will then call the provided Slack webhook? Meaning the architecture actually looks like:

`API Gateway → Lambda router (/slack-to-lambda) → SNS → Command Lambda → API Gateway → Lambda to send message to Slack (/lambda-to-slack)`

If so, what are the benefits of this over the individual commands calling their Slack webhook?


2. > AWS resources for one Slack command are not dependent on the AWS resources for another Slack command.

Can you clarify this? Which resources were being shared in the previous architecture suggestion that they now don’t?


3. > Creating scripts for the Lambda.

By scripts, do you mean the handler function or is it something else?


4. > **Debugging and troubleshooting** - Routing issues can be debugged on the routing repository/router logs. Command related issues can be debugged in the command repository/command logs. 

When I was doing my contribution for Outreachy and I tried making a [version of the Slackbot router](https://github.com/andria-hibe/HOTOSM-Slackbot-Router), the API Gateway already had its own Log Group, and each Lambda function had its own Log Group as well. Is this what you mean?


5. > **Improved maintainability** - Adding a new command will simply involve creating a new GitHub repository. Since the deployment templates of all the lambda functions involve the SNS, Lambda, Logs combination, it is easy to create a new deployment template. Maybe we can even write a template generator as part of this summer.

This seems to be a higher setup cost for each new Lambda function. Additionally, you have to maintain a template generator as well. Can you expand on how this will improve maintainability?

**Comment by arunasank** [comment]
:wave: @andria-hibe thanks for the questions. Here are my replies. Happy to discuss further here, or in our call. cc @ramyaragupathy 

> Regarding the /lambda-to-slack endpoint, I’m not sure how this works. Is this calling another Lambda function that will then call the provided Slack webhook? Meaning the architecture actually looks like:
> 
> `API Gateway → Lambda router (/slack-to-lambda) → SNS → Command Lambda → API Gateway → Lambda to send message to Slack (/lambda-to-slack)`
>
> If so, what are the benefits of this over the individual commands calling their Slack webhook?

* **Separation of codebases** - The biggest advantage is that the code-base for each accompanying command/router is simple, small, and contained to its own repository. This way:
  * The lambda router processes/parses each request and forwards it to the appropriate lambda-SNS pair, and its logic only includes the routing logic. 
  * Each Lambda-SNS repository only has code around the lambda command. The secrets for each command pair is also present in the stack and repository concerned with it.
    Right now we have 5 commands, but assuming we have many more commands in the future, the code base of this stack will become large, even if we separate out the commands into their own folders. Associated documentation and associated secrets will all be part of this stack, which can be a liability. In my past experience, I have seen secrets associated with all commands logged when someone logged the environment variables, to check the value of one secret. Separating out these secrets into their own repositories, as part of the deployment template simplifies this.
* **Ease of new contributions** - If we have a member from another team (Like the tasking manager team), or a member from the community who wants to add a new slack command, they would not have to modify the routing logic or touch the router code. They would just create a new repository, template, and handlers for the new command. This way, adding a new command or maintaining the existing ones is really easy. If we have a template generator that creates a new deployment template, creating the new template is simplified further, since these templates are going to be duplicates of each other.
* **Separation of secrets** - The secrets associated with each lambda function are contained in the deployment template of the associated lambdas.

> Can you clarify this? Which resources were being shared in the previous architecture suggestion that they now don’t?

I think my wording was a little confusing, sorry. To clarify, I meant that the AWS resources are in separate templates. Not sure about serverless, but with CloudFormation, we are able to check which resources are getting modified and which are not based on the deployment script we use now, and confirm the updation of resources if required. Having fewer number of resources per stack makes it easy to confirm/deny the update. This might sound a little confusing, but you can maybe try modifying any of the staging/test stacks using the existing deployment methods to get a sense for what i mean by this.

> By scripts, do you mean the handler function or is it something else?

Handler scripts.

> When I was doing my contribution for Outreachy and I tried making a version of the Slackbot router, the API Gateway already had its own Log Group, and each Lambda function had its own Log Group as well. Is this what you mean?

Yes. :+1: 

> This seems to be a higher setup cost for each new Lambda function. Additionally, we have to maintain a template generator as well. Can you expand on how this will improve maintainability?

* A template generator already exists and is maintained by Mapbox for generating [CloudFormation lambda templates](https://github.com/mapbox/cloudfriend/blob/master/lib/shortcuts/api.md), which we also use at HOT. We could use this, instead of generating new templates. (Sorry remembered this now - a little rusty because I am now doing different things). 
* The setup cost is not higher, it's just separated. 
  * **New architecture**: If we create a new command, we would have to create new files and new templates. Stack deployment will happen using the new template.
  * **Old architecture**: If we create a new command, we would have to update the existing files and templates. Stack deployment would still have to be done on each new command addition, using the old template.

**Comment by andria-hibe** [comment]
Putting this here regarding secrets and how each Lambda will have access to them:

@arunasank talking about the Slackbot router project she was a part of at MapBox:
> When we initially built out the slack-bot-router project, we used a similar architecture to what Andria had suggested earlier, but as we added more and more commands, the stack quickly became very bloated with all of the scripts associated with each command, and with all the docs associated with each command, not to mention that all of the secrets that each command was dealing with was in this stack. 
There were times when someone would be debugging a specific lambda command or looking to add a new lambda command, and print out all the parameters associated with the stack, which would print out all the secrets onto CloudWatch logs leading to a security vulnerability that could only be fixed by rotating every single secret token. 

What I was planning in the initial architecture proposal was that each Lambda would only have access to the secrets it requires (e.g. the /tm-stats project_id' Lambda would only have access to the Slack secret provided by the app to verify the request, whereas the /osmcha_stats hashtag Lambda would have access to the same Slack secret as well as the OSMCha authorization token). With no routing Lambda, each command Lambda would then only have one additional secret (a shared Slack secret).

**Comment by andria-hibe** [comment]
Thanks @arunasank for the answers!

Wanted to clarify this again:
> Regarding the /lambda-to-slack endpoint, I’m not sure how this works. Is this calling another Lambda function that will then call the provided Slack webhook? Meaning the architecture actually looks like: 
`API Gateway → Lambda router (/slack-to-lambda) → SNS → Command Lambda → API Gateway → Lambda to send message to Slack (/lambda-to-slack)`

So does this mean we have an exit proxy Lambda (separate from the first proxy Lambda router) that will receive a message from the command Lambdas using the `/lambda-to-slack` path and will then call the Slack webhook to send a message to Slack?

**Comment by andria-hibe** [comment]
Hi @ramyaragupathy @arunasank,

Before talking to DevOps, can we answer my previous question regarding how we're sending messages back to Slack and how the /lambda-to-slack endpoint will be used?

I would also like to understand more how we will be storing secrets in the CloudFormation deployment template. Can you direct me to any resources on how this will be done? Are the secrets a variable in the template? How are we encrypting it?

**Comment by arunasank** [comment]
> Before talking to DevOps, can we answer my previous question regarding how we're sending messages back to Slack and how the /lambda-to-slack endpoint will be used?

yep. It looks like the Slack response payload contains a `responseUrl` which can be used for all requests, so we don't need the second end point. Just one end point `slack-to-lambda` to forward the requests from Slack to Lambda. :+1: 

> I would also like to understand more how we will be storing secrets in the CloudFormation deployment template. Can you direct me to any resources on how this will be done? Are the secrets a variable in the template? How are we encrypting it?

We use CloudFormation templates at HOT. Secret management on CFN is complicated because there is no clear way to have encrypted CloudFormation parameters. The most common way (and something I have seen work at scale) is to use KMS keys to manage encryption on CloudFormation. The KMS keys are created through CloudFormation again in each region, and a CloudFormation export of the KMS key is used to encrypt parameters in all CloudFormation stacks in the region.

Here's a resource that works using CloudFormation and KMS, works at scale, and uses the same deployment mechanisms as HOT at the moment: https://github.com/mapbox/cloudformation-kms

SSM Parameter Stores would relieve us from creating and maintaining KMS CloudFormation stacks, but unfortunately CloudFormation does not support the "SecureString" SSM Parameter type as of now. The alternative is to create the SSM Parameters manually on the web console, and reference them with CloudFormation parameters. Let's stay away from manual creation of secrets because
* Manual resources are much harder to maintain. If secrets are accidentally deleted/wrong secret is deleted/updated on the console, it can be problematic.
* Manual creation and maintenance of resources will not scale.

**Comment by andria-hibe** [comment]
Thanks for that @arunasank. I have a few questions regarding the implementation which I think are best asked once we're deploying the CloudFormation templates since it's mostly regarding the actual process and I'll need quite a bit of guidance on that. Otherwise, I'm clear on how this architecture flows.

Do you want me to edit the diagram so we can put in the proxy Lambda router which will publish to SNS as well as tweak the part that send the message back to Slack? Then we can use the diagram for the presentation to DevOps? Cheers. @ramyaragupathy

**Comment by ramyaragupathy** [comment]
Thanks for capturing all the discussions @andria-hibe and @arunasank. Here's the final sketch:

![slack-router](https://user-images.githubusercontent.com/12103383/84057200-f6620000-a9d4-11ea-9b74-fd4ecd9eb27b.png)

<details>
<summary>
Outlining the entire flow's manual setup, please feel free to cover details that I have missed here.
</summary>

Input could come from:
* User typed slash commands from Slack
* Github Events

**Slack Application:**

* Setup a Slack app in HOT workspace with `Slash Commands` functionality enabled
* An `OAuth Access Token` is generated for your app
* Setup sufficient write permissions for your Slack app
* Install the app to your workspace

**Set up a router lambda:**

* Create a lambda function with sufficient logic to route to a SNS topic based on the payload
* After identifying the flow of service in the architecture, the code should send a `200` status code to slack with a message if needed - This is needed to keep the workflow alive in our pipeline. We can also send a message to user, to know that the command was successfully received by the app. Without a 200 status code, there is an error message shown for the user.
Per Slack documentation -
`This confirmation must be received by Slack within 3000 milliseconds of the original request being sent, otherwise a Timeout was reached will be displayed to the user.`

* Lambda execution role should have permissions to publish message to SNS Topic

**Setup an API Gateway:**

* Setup a `slack-to-router` HTTP API Gateway with one route set to `POST` method
* Integrate the route to router lambda function - There are two different Lambda integrations you can configure in API Gateway, such as Lambda integration and Lambda proxy integration. For Lambda integration, you can customise what you are going to pass to Lambda in the payload that you don't need to parse the body, but when you are using Lambda Proxy integration in API Gateway, API Gateway will proxy everything to Lambda in payload.

I have tried and tested a normal Lambda integration and it works fine, so we could go with that.
* Publish your API: To publish an API, create an API deployment and associate it with a `stage`. A stage is like a snapshot in time of the API. If we redeploy an API, we can either update an existing stage or create a new one.
On publishing, an invoke URL is generated for the API Gateway which we will use in our Slack Application
* Grant API permissions to invoke lambda function


**Setup a slash command:**

* After you have the basic request-response model in place, add a slash command to the Slack app with API Gateway invoke url as the command's reference URL
* Upon adding a new slash command, app has to be reinstalled to the workspace
* Try this command from Slack and you should get the message from the router lambda - this model should work in public & private channel and direct messages

**Setup the processing lambda:**

* Create a lambda that will be processing the slash command request
* At the end of the processing, this lambda should send a message to Slack using the payload's `response_url`. I used Slack web API
* Here the lambda should have access to Slack app's `OAuth access token`. I got this done with parameter store wherein the value was stored as a secure string with KMS key source
* Using AWS SDK this value can be retrieved and decrypted in the respective processing script


**Setup a SNS topic:**

* Create a SNS topic and set your processing lambda for subscription to this topic
* SNS topic receives the payload from the router lambda and forwards it to the lambda function that's subscribed to it
</details>

**Observations from manual setup:**
* Each service should have attached policies that enables the service role to have sufficient permissions to access subsequent service in the pipeline
* We are not using much secrets or access token related API calls in our workflow, so far I foresee just the Slack app's OAuth Access token
* Even for GitHub events we could use Slack Web API over an incoming webhook, that would save us one additional step on deployment. 

### Github structure and deployment separation

I propose a mono repo for all the slash commands in the following manner:

```
.
├── README.md
├── cloudformation
│   └── slack-router-cfn-template.js
├── package.json
└── src
    ├── README.md
    ├── health-leaderboard
    │   ├── README.md
    │   ├── cloudformation
    │   │   └── health-leaderboard-cfn-template.js
    │   ├── health-leaderboard-lambda.js
    │   └── package.json
    ├── health-tm
    │   ├── README.md
    │   ├── cloudformation
    │   │   └── health-tm-cfn-template.js
    │   ├── health-tm-lambda.js
    │   └── package.json
    ├── label-filter
    │   ├── README.md
    │   ├── cloudformation
    │   │   └── label-filter-cfn-template.js
    │   ├── label-filter-lambda.js
    │   └── package.json
    ├── osmcha-project-hashtags
    │   ├── README.md
    │   ├── cloudformation
    │   │   └── osmcha-project-hashtags-cfn-template.js
    │   ├── osmcha-project-hashtags-lambda.js
    │   └── package.json
    └── tm-stats
        ├── README.md
        ├── cloudformation
        │   └── tm-stats-cfn-template.js
        ├── package.json
        └── tm-stats-lambda.js
```

* Though It is a monorepo, router and the individual slash commands should be deployed independently of each other and the structure is reflective of the same. 
* From the deployment perspective, since the code base will be open for anyone to contribute, whoever wants to add a new command can add their code to the same repo and make a PR, instead of having to wait for one of us from the tech team to create a separate repo for the new command. One repo per command will work well for closed setup where we have a limited set of contributors.
* Considering we're open source - I borrow @arunasank's idea on one lambda per command and @andria-hibe's idea on a monorepo. With these two if we could separate out stack deployments I feel we're good. 

**Initial deployment**
* This would involve running CloudFormation for API gateway and router lambda 
* Slack application setup on workspace will be manual - slash commands and invoke URL has to be created and updated
* Each of our SNS and processing lambda pair will be setup as independent stack and the pair will follow a certain naming convention for easier routing
 
**New command integration at a later stage**
* Contributor makes a PR with scripts and CloudFormation in a new folder
* Ramya/Dakota:
     * squash the PR
     * deploy the new command stack to cloud
     * update the router stack - we could skip this step if our script is generic enough to match slash commands with the SNS topic
     * manually add the slash command to Slack

### Other side notes
* Undecided on how we should go ahead with CloudFormation templates - may be we could use a custom template generator or store a default template on cloud that can be reused with different configuration parameters during deployment
* Noticed that CloudFormation Supports AWS Systems Manager Secure String Parameters in CloudFormation Templates in most of the regions cc @arunasank 
* Slack sends message to API gateway in `application/x-www-form-urlencoded` which means payload will be part of the request URL. Since it is base64 encoded there is an additional step to decode this in the router lambda
* For our slash commands we should use `response_url` from the input payload since it helps post across public/private channels and direct message. For Github events let's go with Slack Web API. In case of Github events, such triggers will be tied up with specific channels and we've the channel ID handy for use in our processing lambda. Instead of creating an incoming webhook for posting message to Slack from Github event, we should stick to Slack Web API since we have the channel information beforehand. Going this route will save one manual step during new command deployment.
* Also for Github event, we now have a `label-filter` lambda that uses the same code to post to multiple channels. Each channel is:
  * tied up to a different set of Github labels
  * has a different incoming webhook for the same lambda
For such lambdas that are tied up with external application triggers - our approach will create one lambda per channel integration eg: `label-filter-starter-issues` will be a different stack from `label-filter-tm-backend` and these two will serve different channels/audiences.
* From AWS documentation, each lambda can handle about 2000 concurrent requests, so if there are multiple users typing in command simultaneously, multiple instances of the lambda will be running in parallel handling the requests.

**Comment by andria-hibe** [comment]
Brilliant! Thank you so much @ramyaragupathy @arunasank for the hard work! 🥳🙏🏽

**Comment by ramyaragupathy** [comment]
Wonderful work on this @andria-hibe and @arunasank 💯
